### PR TITLE
Support environment -> versioned package mapping

### LINKED ISSUE #631: Support environment -> versioned package mapping
/kind feature

Environments should support versioned references to installed packages.
This can be represented by adding `LibraryConfigs` to `EnvironmentConfig`, e.g.

```
environments:
  default:
    destination:
      namespace: default
      server: https://192.168.64.3:8443
    k8sVersion: v1.10.0
    path: default
    libraries:
      nginx:
        name: nginx
        registry: incubator
        version: 40285d8a14f1ac5787e405e1023cf0c07f6aa28c
      spinnaker:
        name: spinnaker
        registry: helm-stable
        version: 2.3.4
```

The global `LibraryConfigs` will still be supported at a lower precedence.

### LINKED ISSUE #651: Package vendor paths should be versioned
As part of #629 and #631 we are adding support for side-by-side versioned instances of a package in our vendor cache. #631 adds support for reading versioned packages, while this issue captures support for writing out correctly versioned directories when installing new packages.

More succinctly:

`ks pkg install incubator/mysql`, where the package is versioned with commit sha `40285d8a14f1ac5787e405e1023cf0c07f6aa28c`, should vendor into the following directory:

`vendor/incubator/mysql@40285d8a14f1ac5787e405e1023cf0c07f6aa28c`

### PR DESCRIPTION
**Pull request by shomron**
This PR adds version-mapped package support to environments. An environment can specify the specific version of a package to consume - and multiple, side-by-side versions of a package can be installed in the package cache.

Additional consistency work should be completed (linked in #629), particularly #651 and #635.

Closes #631 
Closes #651

### CODE REVIEW

**Comment by bryanl** [comment]
@shomron 

With this addition to the code, you shouldn't have to run `ks upgrade` if you run `ks init` after this change. Here is some sample output (I added an error check to avoid a panic in the directory copy code.

```
$ ks show default                             
WARNING Versioned package incubator/apache@40285d8a14f1ac5787e405e1023cf0c07f6aa28c stored in unversioned path - please run `ks upgrade`to correct.  action=pkg.NewLocal
ERROR find objects: revendoring packages for environment: default: copying package /Users/bryan/go/src/github.com/ksonnet/ksonnet/mule2/vendor/incubator/apache@40285d8a14f1ac5787e405e1023cf0c07f6aa28c->/var/folders/b2/nfw6ff8d47xbv8ffx7bs1pl80000gn/T/ksvendor858703045/incubator/apache: walking "/Users/bryan/go/src/github.com/ksonnet/ksonnet/mule2/vendor/incubator/apache@40285d8a14f1ac5787e405e1023cf0c07f6aa28c": lstat /Users/bryan/go/src/github.com/ksonnet/ksonnet/mule2/vendor/incubator/apache@40285d8a14f1ac5787e405e1023cf0c07f6aa28c: no such file or directory
```

File: pkg/util/io/copy_dir.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+)
+
+// CopyRecursive copies the contents of the src directory to the dst directory, recursively.
+// Equivilant to to `cp -R src dst`
+func CopyRecursive(fs afero.Fs, dst string, src string, fileMode os.FileMode, dirMode os.FileMode) error {
+	if fs == nil {
+		return errors.Errorf("nil filesystem interface")
+	}
+
+	return afero.Walk(fs, src, func(path string, info os.FileInfo, err error) error {
```
Comment (bryanl) [inline pkg/util/io/copy_dir.go]: This error needs to be check to ensure everything is fine before starting. If not, there is a chance the `info` var on line 41 will be nil.

Comment (shomron) [reply pkg/util/io/copy_dir.go]: 👍

**Comment by shomron** [comment]
^ @bryanl shall I squash this?

**Comment by bryanl** [comment]
@shomron Yes you should.

**Comment by shomron** [comment]
@bryanl minor fix pushed in https://github.com/ksonnet/ksonnet/pull/652/commits/07852ea3cf23d5f45ab158c71f38ccfc7c86df52 - please have a look.

**Comment by bryanl** [comment]
@shomron I don't see anything there that should hold this work up.

**Pull request merged**